File: shared/schemas/xai.py
# shared/schemas/xai.py
from typing import List, Dict, Any, Optional, Tuple
from pydantic import BaseModel, ConfigDict, Field

# --- Schemas for XAI result_data structures ---

class FilePredictionDetail(BaseModel):
    """Detailed prediction result for a single file/class instance."""
    file: Optional[str] = Field(None, description="File path.")
    class_name: Optional[str] = Field(None, alias="class", description="Class name (if applicable).") # Use alias if frontend expects 'class'
    prediction: int = Field(..., description="Predicted label (e.g., 0 or 1) for this instance.")
    probability: float = Field(..., description="Probability of the positive class (defect-prone) for this instance.")

    model_config = ConfigDict(populate_by_name=True) # Allow using alias 'class'

class FeatureImportanceValue(BaseModel):
    feature: str = Field(..., description="Name of the feature.")
    importance: float = Field(..., description="Calculated importance score (e.g., mean absolute SHAP).")

class FeatureImportanceResultData(BaseModel):
    feature_importances: List[FeatureImportanceValue] = Field(..., description="List of feature importance scores.")

class FeatureSHAPValue(BaseModel):
    feature: str = Field(..., description="Name of the feature.")
    value: float = Field(..., description="The calculated SHAP value for this feature and instance.")
    feature_value: Any = Field(None, description="The actual value of the feature for this instance (optional).")

class InstanceSHAPResult(BaseModel):
    file: Optional[str] = Field(None, description="File path associated with this instance.")
    class_name: Optional[str] = Field(None, alias="class", description="Class name associated with this instance (if applicable).")
    shap_values: List[FeatureSHAPValue] = Field(..., description="List of SHAP values per feature for this instance.")
    base_value: Optional[float] = Field(None, description="SHAP base value for this instance's prediction.")

    model_config = {"populate_by_name": True}

class SHAPResultData(BaseModel):
    instance_shap_values: List[InstanceSHAPResult] = Field(..., description="List containing SHAP value details for each analyzed file/class instance.")

class InstanceLIMEResult(BaseModel):
    file: Optional[str] = Field(None, description="File path.")
    class_name: Optional[str] = Field(None, alias="class", description="Class name.")
    explanation: List[Tuple[str, float]] = Field(..., description="List of (feature, weight) tuples from LIME.")

    model_config = {"populate_by_name": True}

class LIMEResultData(BaseModel):
    instance_lime_values: List[InstanceLIMEResult] = Field(..., description="List of LIME explanations per instance.")

class CounterfactualExample(BaseModel):
     features: Dict[str, Any] = Field(..., description="The counterfactual feature values.")
     outcome_probability: float = Field(..., description="Predicted probability with these counterfactual features.")

class InstanceCounterfactualResult(BaseModel):
    file: Optional[str] = Field(None, description="File path.")
    class_name: Optional[str] = Field(None, alias="class", description="Class name.")
    counterfactuals: List[CounterfactualExample] = Field(..., description="List of counterfactual examples found.")

    model_config = {"populate_by_name": True}

class CounterfactualResultData(BaseModel):
    instance_counterfactuals: List[InstanceCounterfactualResult] = Field(...)

class DecisionPathNode(BaseModel):
    id: str
    condition: Optional[str] = None # None for leaf nodes
    samples: Optional[int] = None
    # Value might represent class probabilities or counts
    value: Optional[List[float]] | Optional[List[int]] = None

class DecisionPathEdge(BaseModel):
    source: str
    target: str
    label: Optional[str] = None # e.g., "True" / "False" or feature value threshold

class InstanceDecisionPath(BaseModel):
    file: Optional[str] = Field(None, description="File path.")
    class_name: Optional[str] = Field(None, alias="class", description="Class name.")
    nodes: List[DecisionPathNode]
    edges: List[DecisionPathEdge]

    model_config = {"populate_by_name": True}

class DecisionPathResultData(BaseModel):
    instance_decision_paths: List[InstanceDecisionPath] = Field(..., description="Decision path details for each applicable instance.")

File: frontend/types/api/xai.ts
// frontend/types/api/xai.ts
import { XAITypeEnum, XAIStatusEnum } from "./enums";

// --- Schemas from shared/schemas/xai.py ---
// Note: Make sure field names match the Pydantic models, especially if using `alias` in Pydantic.
// For example, `class_name: Optional[str] = Field(None, alias="class", ...)` means frontend might receive "class".
// However, for consistency and to avoid issues with TypeScript property names, using `class_name` is often preferred
// and the backend should ideally serialize with `by_alias=False` or ensure frontend matches the alias.
// Assuming backend sends `class_name` or `populate_by_name=True` is used carefully in Pydantic.

export interface FilePredictionDetail { // This is also used by InferenceJob, so it needs to be consistent or imported.
  file_path: string; // Path to the file
  predicted_label: number; // The prediction for this specific file
  prediction_probability: number; // The probability for this prediction
  class_name?: string | null; // Optional class name if applicable
  // file?: string; // Alternative if schema uses 'file' instead of 'file_path'
}

export interface FeatureImportanceValue {
  feature: string;
  importance: number;
}

export interface FeatureImportanceResultData {
  feature_importances: FeatureImportanceValue[];
}

export interface FeatureSHAPValue {
  feature: string;
  value: number;
  feature_value?: any;
}

export interface InstanceSHAPResult {
  file?: string | null;
  class_name?: string | null; // Using class_name instead of 'class'
  shap_values: FeatureSHAPValue[];
  base_value?: number;
}

export interface SHAPResultData {
  instance_shap_values: InstanceSHAPResult[];
}

export interface InstanceLIMEResult {
  file?: string | null;
  class_name?: string | null; // Using class_name
  explanation: [string, number][]; // List of (feature, weight) tuples
}

export interface LIMEResultData {
  instance_lime_values: InstanceLIMEResult[];
}

export interface CounterfactualExample {
  features: Record<string, any>;
  outcome_probability: number;
}

export interface InstanceCounterfactualResult {
  file?: string | null;
  class_name?: string | null; // Using class_name
  counterfactuals: CounterfactualExample[];
}

export interface CounterfactualResultData {
  instance_counterfactuals: InstanceCounterfactualResult[];
}

export interface DecisionPathNode {
  id: string; // Node ID
  condition?: string | null; // Condition for splitting (non-leaf)
  samples?: number | null; // Number of samples at this node
  value?: number[] | number[][] | null; // Class distribution or value at leaf
}

export interface DecisionPathEdge {
  source: string; // Source node ID
  target: string; // Target node ID
  label?: string | null; // Label for the edge (e.g., "True", "False")
}

export interface InstanceDecisionPath {
  file?: string | null;
  class_name?: string | null; // Using class_name
  nodes: DecisionPathNode[];
  edges: DecisionPathEdge[];
}

export interface DecisionPathResultData {
  instance_decision_paths: InstanceDecisionPath[];
}

// --- Schemas from shared/schemas/xai_job.py ---
export interface XAIResultRead {
  id: number;
  inference_job_id: number;
  xai_type: XAITypeEnum;
  status: XAIStatusEnum;
  status_message?: string | null;
  result_data?: FeatureImportanceResultData | SHAPResultData | LIMEResultData | CounterfactualResultData | DecisionPathResultData | Record<string, any> | null;
  celery_task_id?: string | null;
  started_at?: string | null; // ISO date string
  completed_at?: string | null; // ISO date string
  created_at: string; // ISO date string
  updated_at: string; // ISO date string
}

export interface XAITriggerResponse {
  task_id?: string | null;
  message: string;
}
